### PR TITLE
Add option to keep files in S3 when publishing

### PR DESCRIPTION
**Pull request by ghing**
Add a '--keep-files' option to the publish management command
that causes the command to skip the deletion of files that
have keys in the S3 bucket, but don't exist in the build directory.

This is a simple way to support safer publishing in cases where
items may be written to the S3 bucket outside of the Django
Bakery build/publish workflow.  With this option, users can
publish their Django project with Django Bakery and have the
flexibility to manage other components that publish to the
same S3 bucket outside of Django.

To test this feature, use [moto](https://github.com/spulec/moto)
to mock the boto library, used to interact with S3.  The mock
package is also needed to work around some weirdness with the
S3 mocking that isn't crucial to the feature under test.

Update setup.py to include the testing dependencies.

### CODE REVIEW

**Comment by ghing** [comment]
It looks like tests_require isn't getting processed when running ``setup.py test`` perhaps because of https://bitbucket.org/pypa/setuptools/issue/232/overriding-commands-is-inconsistent-and

**Comment by palewire** [comment]
This is a great idea. I might not have time to give it a thorough read today but I'm excited to get to it!

**Comment by palewire** [comment]
This looks great to me, though I wonder (pedantically) if there might be a more precise name for the option than ``keep-files``. My slight concern is that it might be a little unclear what "files" the flag is referring to. Could there be something more explicit we could say like "do-not-delete"? 

I wonder what term other libraries use in this circumstance. I see that s3cmd does not delete by default and requires a ``delete-removed`` flag, a system I like even less.

What do you think?

**Comment by ghing** [comment]
@palewire I agree.  I think ``do-not-delete``, or the shorter ``no-delete`` is more descriptive of what's actually going on.

In retrospect, I think my preferred behavior would be to not delete by default and have to specify the a flag to enable this behavior, though s3cmd's ``delete-removed`` name is less than terrific.  My big motivation for this is to follow the behavior of the venerable ``rsync``, which doesn't delete unless a ``delete`` flag is specified.  I'll also note that the official ``aws`` CLI utility doesn't provide any way to toggle this behavior.  I think it doesn't  delete. Ultimately, however, I'm not going to advocate changing the default behavior for a mature project.  

Also, interesting, in [this list of GNU options](https://www.gnu.org/prep/standards/html_node/Option-Table.html#Option-Table), ``csplit`` has a ``keep-files`` option which has a similar meaning (but without the sync context) and ``no-*`` seems common for disabling functionality.

In short, my vote is for ``no-delete``.

**Comment by palewire** [comment]
I'm sold. Let's roll with ``no-delete``.

**Comment by palewire** [comment]
You are in master. I cut out the new mock tests for now just because I don't have time to study them and I want to get your excellent option into the master branch as fast as possible. Let's pick up the test work soon in #21.
